diff --git a/src/OpenFOAM/meshes/meshShapes/cellShape/cellShape.H b/src/OpenFOAM/meshes/meshShapes/cellShape/cellShape.H
index 1915de8..4be8a88 100644
--- a/src/OpenFOAM/meshes/meshShapes/cellShape/cellShape.H
+++ b/src/OpenFOAM/meshes/meshShapes/cellShape/cellShape.H
@@ -103,7 +103,8 @@ public:
 
         //- Return the points corresponding to this cellShape
         inline pointField points(const pointField& meshPoints) const;
-
+        inline labelList pointsLabel() const;
+        
         //- Model reference
         inline const cellModel& model() const;
 
diff --git a/src/OpenFOAM/meshes/meshShapes/cellShape/cellShapeI.H b/src/OpenFOAM/meshes/meshShapes/cellShape/cellShapeI.H
index 6716231..90fe619 100644
--- a/src/OpenFOAM/meshes/meshShapes/cellShape/cellShapeI.H
+++ b/src/OpenFOAM/meshes/meshShapes/cellShape/cellShapeI.H
@@ -85,6 +85,22 @@ inline Foam::pointField Foam::cellShape::points
 }
 
 
+inline Foam::labelList Foam::cellShape::pointsLabel() const
+{
+    // There are as many points as there labels for them
+    labelList p(size());
+
+    // For each point in list, set it to the point in 'pnts' addressed
+    // by 'labs'
+    forAll(p, i)
+    {
+        p[i] = operator[](i);
+    }
+
+    // Return list
+    return p;
+}
+
 inline const Foam::cellModel& Foam::cellShape::model() const
 {
     return *m;
diff --git a/src/OpenFOAM/meshes/meshShapes/face/face.H b/src/OpenFOAM/meshes/meshShapes/face/face.H
index c481258..9949475 100644
--- a/src/OpenFOAM/meshes/meshShapes/face/face.H
+++ b/src/OpenFOAM/meshes/meshShapes/face/face.H
@@ -180,6 +180,7 @@ public:
 
         //- Return the points corresponding to this face
         inline pointField points(const pointField&) const;
+        inline labelList pointsLabel() const;
 
         //- Centre point of face
         point centre(const pointField&) const;
diff --git a/src/OpenFOAM/meshes/meshShapes/face/faceI.H b/src/OpenFOAM/meshes/meshShapes/face/faceI.H
index 7dfc9ef..e1bf1b1 100644
--- a/src/OpenFOAM/meshes/meshShapes/face/faceI.H
+++ b/src/OpenFOAM/meshes/meshShapes/face/faceI.H
@@ -93,6 +93,22 @@ inline Foam::pointField Foam::face::points(const pointField& meshPoints) const
     return p;
 }
 
+inline Foam::labelList Foam::face::pointsLabel() const
+{
+    // There are as many points as there labels for them
+    labelList p(size());
+
+    // For each point in list, set it to the point in 'pnts' addressed
+    // by 'labs'
+    forAll(p, i)
+    {
+        p[i] = operator[](i);
+    }
+
+    // Return list
+    return p;
+}
+
 
 inline Foam::scalar Foam::face::mag(const pointField& p) const
 {
diff --git a/src/mesh/blockMesh/blockMesh/blockMesh.H b/src/mesh/blockMesh/blockMesh/blockMesh.H
index 0188d28..35d7a26 100644
--- a/src/mesh/blockMesh/blockMesh/blockMesh.H
+++ b/src/mesh/blockMesh/blockMesh/blockMesh.H
@@ -203,6 +203,9 @@ public:
 
             //- Enable/disable verbose information about the progress
             static void verbose(const bool on=true);
+            
+            //- Change point location (smoothing
+            void setPoints(const pointField &point){points_ = point;}
 
         // Write
 
